# *HEART DISEASE PREDICTION*

In [34]:
import pandas as pd
#loading the dataset
dataset=pd.read_csv("heart_disease_prediction.csv")

In [35]:
#printing the first and last 5 rows of the dataset
dataset

Age  Sex  Chest pain type   BP  Cholesterol  FBS over 120  EKG results  \
0     70    1                4  130          322             0            2   
1     67    0                3  115          564             0            2   
2     57    1                2  124          261             0            0   
3     64    1                4  128          263             0            0   
4     74    0                2  120          269             0            2   
..   ...  ...              ...  ...          ...           ...          ...   
265   52    1                3  172          199             1            0   
266   44    1                2  120          263             0            0   
267   56    0                2  140          294             0            2   
268   57    1                4  140          192             0            0   
269   67    1                4  160          286             0            2   

     Max HR  Exercise angina  ST depression  Slope of ST  \
0       109                0            2.4            2   
1       160                0            1.6            2   
2       141                0            0.3            1   
3       105                1            0.2            2   
4       121                1            0.2            1   
..      ...              ...            ...          ...   
265     162                0            0.5            1   
266     173                0            0.0            1   
267     153                0            1.3            2   
268     148                0            0.4            2   
269     108                1            1.5            2   

     Number of vessels fluro  Thallium Heart Disease  
0                          3         3      Presence  
1                          0         7       Absence  
2                          0         7      Presence  
3                          1         7       Absence  
4                          1         3       Absence  
..                       ...       ...           ...  
265                        0         7       Absence  
266                        0         7       Absence  
267                        0         3       Absence  
268                        0         6       Absence  
269                        3         3      Presence  

[270 rows x 14 columns]

In [36]:
dataset.info() #shows the datatype,count,coulmns,memory usage etc..
dataset.shape 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 270 entries, 0 to 269
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      270 non-null    int64  
 1   Sex                      270 non-null    int64  
 2   Chest pain type          270 non-null    int64  
 3   BP                       270 non-null    int64  
 4   Cholesterol              270 non-null    int64  
 5   FBS over 120             270 non-null    int64  
 6   EKG results              270 non-null    int64  
 7   Max HR                   270 non-null    int64  
 8   Exercise angina          270 non-null    int64  
 9   ST depression            270 non-null    float64
 10  Slope of ST              270 non-null    int64  
 11  Number of vessels fluro  270 non-null    int64  
 12  Thallium                 270 non-null    int64  
 13  Heart Disease            270 non-null    object 
dtypes: float64(1), int64(12), 

(270, 14)

There are no null values in the dataset.

In [37]:
dataset.describe()

Age         Sex  Chest pain type          BP  Cholesterol  \
count  270.000000  270.000000       270.000000  270.000000   270.000000   
mean    54.433333    0.677778         3.174074  131.344444   249.659259   
std      9.109067    0.468195         0.950090   17.861608    51.686237   
min     29.000000    0.000000         1.000000   94.000000   126.000000   
25%     48.000000    0.000000         3.000000  120.000000   213.000000   
50%     55.000000    1.000000         3.000000  130.000000   245.000000   
75%     61.000000    1.000000         4.000000  140.000000   280.000000   
max     77.000000    1.000000         4.000000  200.000000   564.000000   

       FBS over 120  EKG results      Max HR  Exercise angina  ST depression  \
count    270.000000   270.000000  270.000000       270.000000      270.00000   
mean       0.148148     1.022222  149.677778         0.329630        1.05000   
std        0.355906     0.997891   23.165717         0.470952        1.14521   
min        0.000000     0.000000   71.000000         0.000000        0.00000   
25%        0.000000     0.000000  133.000000         0.000000        0.00000   
50%        0.000000     2.000000  153.500000         0.000000        0.80000   
75%        0.000000     2.000000  166.000000         1.000000        1.60000   
max        1.000000     2.000000  202.000000         1.000000        6.20000   

       Slope of ST  Number of vessels fluro    Thallium  
count   270.000000               270.000000  270.000000  
mean      1.585185                 0.670370    4.696296  
std       0.614390                 0.943896    1.940659  
min       1.000000                 0.000000    3.000000  
25%       1.000000                 0.000000    3.000000  
50%       2.000000                 0.000000    3.000000  
75%       2.000000                 1.000000    7.000000  
max       3.000000                 3.000000    7.000000

In [38]:
dataset.columns #getting the names of the columns

Index(['Age', 'Sex', 'Chest pain type', 'BP', 'Cholesterol', 'FBS over 120',
       'EKG results', 'Max HR', 'Exercise angina', 'ST depression',
       'Slope of ST', 'Number of vessels fluro', 'Thallium', 'Heart Disease'],
      dtype='object')

In [39]:
#checking target distribution
dataset['Heart Disease'].value_counts()

Absence     150
Presence    120
Name: Heart Disease, dtype: int64

In [40]:
#splitting target and features
x=dataset.drop(columns='Heart Disease',axis=1)
y=dataset['Heart Disease']

In [41]:
#splitting into training and testing data
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,stratify=y,random_state=2)

In [42]:
print(x.shape,x_train.shape,x_test.shape)

(270, 13) (216, 13) (54, 13)


*The data is split into 80% and 20% as train data and test data.*

In [43]:
#training the model
from sklearn.linear_model import LogisticRegression
model=LogisticRegression(solver='lbfgs',max_iter=1000)

In [44]:
model.fit(x_train.values,y_train.values)

LogisticRegression(max_iter=1000)

In [45]:
#evaluation of performance
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
x_train_prediction = model.predict(x_train.values)
training_data_accuracy = accuracy_score(x_train_prediction,y_train.values)
print('accuracy of training data:',training_data_accuracy)

accuracy of training data: 0.8888888888888888


In [46]:
x_test_prediction = model.predict(x_test.values)
testing_data_accuracy = accuracy_score(x_test_prediction,y_test.values)
print('accuracy of testing data:',testing_data_accuracy)

accuracy of testing data: 0.7962962962962963


In [47]:
confusion=confusion_matrix(y_test.values,x_test_prediction)
print('confusion matrix is:',confusion)

confusion matrix is: [[25  5]
 [ 6 18]]


In [48]:
report=classification_report(y_test.values,x_test_prediction)
print('classifiction report is:',report)

classifiction report is:               precision    recall  f1-score   support

     Absence       0.81      0.83      0.82        30
    Presence       0.78      0.75      0.77        24

    accuracy                           0.80        54
   macro avg       0.79      0.79      0.79        54
weighted avg       0.80      0.80      0.80        54



* accuracy of the model is 80%,meaning it correctly predicted the presence or absence of heart disease in 80% of the cases.
* The precision for predicting absence is higher (0.81) than for predicting presence (0.78), indicating that the model is slightly more reliable in predicting patients without heart disease.
* The recall for predicting presence is lower (0.75) than for absence (0.83), suggesting that there were more false negatives in predicting heart disease.
* The F1-score, which balances precision and recall, is also higher for absence predictions.

for example:
enter the values of 'Age', 'Sex', 'Chest pain type', 'BP', 'Cholesterol', 'FBS over 120',
       'EKG results', 'Max HR', 'Exercise angina', 'ST depression',
       'Slope of ST', 'Number of vessels fluro', 'Thallium' respectively in the input_data

In [49]:
import numpy as np
#predictive system building
input_data=(51,0,4,130,305,0,0,142,1,1.2,2,0,7)
input_data_asarray=np.array(input_data)
input_reshaped=input_data_asarray.reshape(1,-1)
prediction=model.predict(input_reshaped)
print('As per the prediction by the model:')
if prediction[0]=='Presence':
    print('The Person has a Heart Disease')
else:
    print('The Person doesnot have a Heart Disease')

As per the prediction by the model:
The Person has a Heart Disease
